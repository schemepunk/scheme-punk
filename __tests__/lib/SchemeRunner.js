'use strict';

const _ = require('lodash');

const SchemeRunner = require('../../lib/SchemeRunner');
const SchemePunkErrors = require('../../lib/SchemePunkErrors');


let tmpMocks = [];
let data;
let schemeArray;
let molotovOptions;

beforeEach(() => {
  tmpMocks.forEach(mock => mock.mockRestore());
  tmpMocks = [];
  jest.resetAllMocks();
  jest.spyOn(Date, 'now').mockReturnValue(2000);
  data = {test: 'testValue'};
  schemeArray = {
    arbitraryNameForThisScheme: [
      [
        {
          source: {
            target: 'test',
            plugin: 'originalSchemeSource'
          },
          transform: {
            plugin: 'prependValuesAdapter',
            sourcePrepend: '+'
          },
          destination: {
            target: 'test'
          }
        }
      ],
      [
        {
          source: {
            target: 'test',
            plugin: 'originalSchemeSource'
          },
          transform: {
            plugin: 'appendValuesAdapter',
            sourceAppend: '-'
          },
          destination: {
            target: 'test',
            plugin: 'concatIntoDestination'
          }
        }
      ]
    ],
  };
  molotovOptions = {
    overrides: {},
    cocktailClasses: []
  };
});

afterAll(() => {
  jest.restoreAllMocks();
});

describe('Scheme Runner', () => {
  test('Basic init and scheme running', () => {
    expect.assertions(5);
    const schemeRunner = new SchemeRunner();
    expect(schemeRunner).toBeInstanceOf(SchemeRunner);
    return schemeRunner.init(_.cloneDeep(data), _.cloneDeep(schemeArray), _.cloneDeep(molotovOptions))
      .then((sr) => {
        expect(sr.getData()).toMatchSnapshot();
        expect(sr.getSchemes()).toMatchSnapshot();
        expect(sr.getMolotovOptions()).toMatchSnapshot();
        return schemeRunner.runScheme();
      })
      .then(item => expect(item).toEqual({
        test: [
          '+testValue',
          'testValue-'
        ]
      }));
  });
});

describe('Scheme Runner Error Cases', () => {
  test('No data', () => {
    expect.assertions(1);
    const schemeRunner = new SchemeRunner();
    return schemeRunner.init(undefined, _.cloneDeep(schemeArray), _.cloneDeep(molotovOptions))
      .catch(e => expect(e).toBeInstanceOf(SchemePunkErrors));
  });
  test('No Scheme', () => {
    expect.assertions(1);
    const schemeRunner = new SchemeRunner();
    return schemeRunner.init(_.cloneDeep(data), undefined, _.cloneDeep(molotovOptions))
      .catch(e => expect(e).toBeInstanceOf(SchemePunkErrors));
  });
});


describe('Scheme Runner Coverage', () => {
  test('Basic init and scheme running', () => {
    expect.assertions(5);
    const schemeRunner = new SchemeRunner();
    expect(schemeRunner).toBeInstanceOf(SchemeRunner);
    const tmpData = {
      originalScheme: _.cloneDeep(data),
      activeScheme: _.cloneDeep(data),
      newScheme: _.cloneDeep(data)
    };
    return schemeRunner.init(tmpData, _.cloneDeep(schemeArray), _.cloneDeep(molotovOptions))
      .then((sr) => {
        expect(sr.getData()).toMatchSnapshot();
        expect(sr.getSchemes()).toMatchSnapshot();
        expect(sr.getMolotovOptions()).toMatchSnapshot();
        return schemeRunner.runScheme();
      })
      .then(item => expect(item).toEqual({
        test: [
          '+testValue',
          'testValue-'
        ]
      }));
  });
});
