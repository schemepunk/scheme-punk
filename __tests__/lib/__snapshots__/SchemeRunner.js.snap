// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scheme Runner Basic init and scheme running 1`] = `
Object {
  "activeScheme": Object {
    "test": "testValue",
  },
  "newScheme": Object {
    "test": "testValue",
  },
  "originalScheme": Object {
    "test": "testValue",
  },
}
`;

exports[`Scheme Runner Basic init and scheme running 2`] = `
Object {
  "arbitraryNameForThisScheme": Array [
    Array [
      Object {
        "destination": Object {
          "target": "test",
        },
        "source": Object {
          "plugin": "originalSchemeSource",
          "target": "test",
        },
        "transform": Object {
          "plugin": "prependValuesAdapter",
          "sourcePrepend": "+",
        },
      },
    ],
    Array [
      Object {
        "destination": Object {
          "plugin": "concatIntoDestination",
          "target": "test",
        },
        "source": Object {
          "plugin": "originalSchemeSource",
          "target": "test",
        },
        "transform": Object {
          "plugin": "appendValuesAdapter",
          "sourceAppend": "-",
        },
      },
    ],
  ],
}
`;

exports[`Scheme Runner Basic init and scheme running 3`] = `
Object {
  "cocktailClasses": Array [],
  "overrides": Object {},
}
`;

exports[`Scheme Runner Coverage Basic init and scheme running 1`] = `
Object {
  "activeScheme": Object {
    "test": "testValue",
  },
  "newScheme": Object {
    "test": "testValue",
  },
  "originalScheme": Object {
    "test": "testValue",
  },
}
`;

exports[`Scheme Runner Coverage Basic init and scheme running 2`] = `
Object {
  "arbitraryNameForThisScheme": Array [
    Array [
      Object {
        "destination": Object {
          "target": "test",
        },
        "source": Object {
          "plugin": "originalSchemeSource",
          "target": "test",
        },
        "transform": Object {
          "plugin": "prependValuesAdapter",
          "sourcePrepend": "+",
        },
      },
    ],
    Array [
      Object {
        "destination": Object {
          "plugin": "concatIntoDestination",
          "target": "test",
        },
        "source": Object {
          "plugin": "originalSchemeSource",
          "target": "test",
        },
        "transform": Object {
          "plugin": "appendValuesAdapter",
          "sourceAppend": "-",
        },
      },
    ],
  ],
}
`;

exports[`Scheme Runner Coverage Basic init and scheme running 3`] = `
Object {
  "cocktailClasses": Array [],
  "overrides": Object {},
}
`;
